extensions/guacamole-auth-restrict/src/main/java/org/apache/guacamole/auth/restrict/user/RestrictedUser.java

```diff
@@ -0,0 +1,184 @@
+    
+    /**
+     * The list of all user attributes provided by this User implementation.
+     */
+    public static final List<String> RESTRICT_USER_ATTRIBUTES = Arrays.asList(
+            RESTRICT_TIME_ALLOWED_ATTRIBUTE_NAME,
+            RESTRICT_TIME_DENIED_ATTRIBUTE_NAME,
+            RESTRICT_HOSTS_ALLOWED_ATTRIBUTE_NAME,
+            RESTRICT_HOSTS_DENIED_ATTRIBUTE_NAME
+    );
+    
+    /**
+     * The form containing the list of fields for the attributes provided
+     * by this module.
+     */
+    public static final Form RESTRICT_LOGIN_FORM = new Form("restrict-login-form",
+            Arrays.asList(
+                    new TimeRestrictionField(RESTRICT_TIME_ALLOWED_ATTRIBUTE_NAME),
+                    new TimeRestrictionField(RESTRICT_TIME_DENIED_ATTRIBUTE_NAME),
+                    new HostRestrictionField(RESTRICT_HOSTS_ALLOWED_ATTRIBUTE_NAME),
+                    new HostRestrictionField(RESTRICT_HOSTS_DENIED_ATTRIBUTE_NAME)
+            )
+    );
+    
+    /**
+     * Wraps the given User object, providing capability of further restricting
+     * logins beyond the default restrictions provided by default modules.
+     *
+     * @param user
+     *     The User object to wrap.
+     * 
+     * @param remoteAddress
+     *     The remote address of the client from which the current user is logged
+     *     in.
+     */
+    public RestrictedUser(User user, String remoteAddress) {
+        super(user);
+        this.remoteAddress = remoteAddress;
+    }
+
+    /**
+     * Returns the User object wrapped by this RestrictUser.
+     *
+     * @return
+     *     The wrapped User object.
+     */
+    public User getUndecorated() {
+        return getDelegateUser();
+    }
+
+    @Override
+    public Map<String, String> getAttributes() {
+
+        // Create independent, mutable copy of attributes
+        Map<String, String> attributes = new HashMap<>(super.getAttributes());
+        
+        // Loop through extension-specific attributes, adding ones that are
+        // empty so that they are displayed in the web UI.
+        for (String attribute : RESTRICT_USER_ATTRIBUTES) {
+            String value = attributes.get(attribute);
+            if (value == null || value.isEmpty())
+                attributes.put(attribute,  null);
+        }
+
+        return attributes;
+
+    }
+
+    @Override
+    public void setAttributes(Map<String, String> attributes) {
+
+        // Create independent, mutable copy of attributes
+        attributes = new HashMap<>(attributes);
+
+        // Loop through extension-specific attributes, only sending ones
+        // that are non-null and non-empty to the underlying storage mechanism.
+        for (String attribute : RESTRICT_USER_ATTRIBUTES) {
+            String value = attributes.get(attribute);
+            if (value != null && value.isEmpty())
+                attributes.put(attribute, null);
+        }
+
+        super.setAttributes(attributes);
```

Comment:
Depending on context, the attributes controlling user restrictions will need to be removed from the `Map` before passing things along to `super.setAttributes()`, to ensure users cannot manipulate their own restrictions.

Reply:
Hmmm, I guess I thought the underlying storage mechanism (that this extension decorates) would enforce the correct permissions, but I've added the admin check in to make sure that only admins can modify user-level restrictions.